Migrate set-related predicates to new design

JSS95 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Succeeds #20656

#### Brief description of what is fixed or changed

This PR migrates the part of old-design predicates to new design.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

JSS95 commented:
Ping to the reviewers of previous PR - @oscarbenjamin @eric-wieser. This PR migrates the predicates for assumption keys such as `real`, `imaginary` to new predicate design. If this looks okay, then I'll do the same for other predicates in next PRs.

oscarbenjamin commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -31,682 +38,753 @@ def _number(expr, assumptions):
+def _(expr, assumptions):
+    return expr.q != 0
+
+@AlgebraicHandler.register(asin)
+@AlgebraicHandler.register(atan)
+@AlgebraicHandler.register(cos)
+@AlgebraicHandler.register(exp)
+@AlgebraicHandler.register(sin)
+@AlgebraicHandler.register(tan)
+def _(expr, assumptions):
```
Perhaps it could be made possible to pass a list to `register` or perhaps there could be a `register_many`

JSS95 replied:
Good idea. Or maybe it would be better to use loop like
```python
for cls in (asin, atan, ...):
    AlgebraicHandler.register(func)
```

oscarbenjamin replied:
There are lots of loops already. I prefer the decorator style but it could be made more compact for cases like this.

JSS95 replied:
I implemented `register_many` method.

JSS95 committed:
Migrate set-related predicates to new Predicate design

oscarbenjamin commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -31,682 +33,650 @@ def _number(expr, assumptions):
+
+for cls in (Exp1, GoldenRatio, ImaginaryUnit, Infinity, NaN, NegativeInfinity,
+            Pi, Rational, TribonacciConstant):
+    IntegerPredicate.register(cls)(lambda expr, assumptions: False)
```
Can these sorts of examples not use `register_many`?

JSS95 replied:
Oh, right. I overlook them...

eric-wieser commented on sympy/assumptions/assume.py:
```diff
@@ -291,11 +292,27 @@ def name(self):
+        """
+        def _(func):
+            for t in types:
+                if is_sequence(t):
+                    cls.register(*t, **kwargs)(func)
+                else:
+                    cls.register(t, **kwargs)(func)
```
```suggestion
                if not is_sequence(t):
                    t = (t,)  # for convience, allow passing `type` to mean `(type,)`
                cls.register(*t, **kwargs)(func)
```

JSS95 committed:
Update sympy/assumptions/assume.py

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

JSS95 committed:
Remove register using loop

JSS95 committed:
Merge branch 'migrate_sethandlers' of https://github.com/JSS95/sympy into migrate_sethandlers

oscarbenjamin commented:
This looks good to me although I'm unsure what it does to the docs. I see you've removed docstrings from the properties on `Q` and added them to the Predicate subclasses which I think makes sense. Do they then show up in the docs somewhere else?

I can't see where those docstrings show up anyway. The handler docs are pretty bad in html:
https://docs.sympy.org/latest/modules/assumptions/handlers/sets.html

JSS95 commented:
@oscarbenjamin 

I think we need `doc/src/modules/assumptions/predicates` page to show predicate docstring.

JSS95 committed:
Add predicates.rst to docs

JSS95 committed:
Add predicates.rst to toctree

JSS95 committed:
Improve the doc of handlers

JSS95 commented:
It seems that `help` on predicate shows the docstring of handler. Here is the result from `help(Q.integer)`. You can see the description of handler appears in attributes section.

```
class IntegerPredicate(sympy.assumptions.assume.Predicate)
 |  IntegerPredicate(*args, **kwargs)
 |  
 |  Integer predicate.
 |  
 |  Explanation
 |  ===========
 |  
 |  ``Q.integer(x)`` is true iff ``x`` belongs to the set of integer
 |  numbers.
 |  
 |  Examples
 |  ========
 |  
 |  >>> from sympy import Q, ask, S
 |  >>> ask(Q.integer(5))
 |  True
 |  >>> ask(Q.integer(S(1)/2))
 |  False
 |  
 |  References
 |  ==========
 |  
 |  .. [1] https://en.wikipedia.org/wiki/Integer
 |  
 |  Method resolution order:
 |      IntegerPredicate
 |      sympy.assumptions.assume.Predicate
 |      sympy.logic.boolalg.Boolean
 |      sympy.core.basic.Basic
 |      sympy.printing.defaults.Printable
 |      builtins.object
 |  
 |  Readonly properties defined here:
 |  
 |  __sympy__
 |  
 |  ----------------------------------------------------------------------
 |  Data and other attributes defined here:
 |  
 |  default_assumptions = {}
 |  
 |  handler = <dispatched IntegerHandler>
 |      Multiply dispatched method: IntegerHandler
 |      
 |      Handler for Q.integer.
 |      
 |      Test that an expression belongs to the field of integer numbers.
 |      
 |      Inputs: <Mul>
 |      --------------
 |      Integer*Integer      -> Integer
 |          Integer*Irrational   -> !Integer
 |          Odd/Even             -> !Integer
 ...
```

JSS95 committed:
Enhance docstring of Predicate

Modified PredicateMeta to append the docstring of handlers to
the class docstring.

JSS95 committed:
Implement PredicateMeta.__doc__

JSS95 committed:
Remove docstring of PredicateMeta

JSS95 committed:
Enhance PredicateMeta.__doc__

JSS95 committed:
Modify documentations more

JSS95 committed:
Fix test failure and enhance __doc__

JSS95 commented:
Okay, I managed to modify `__doc__` so that the handler is properly shown in both `help` and documentation.

Here is the result from `help(Q.integer)`

```
class IntegerPredicate(sympy.assumptions.assume.Predicate)
 |  IntegerPredicate(*args, **kwargs)
 |  
 |  Integer predicate.
 |  
 |  Explanation
 |  ===========
 |  
 |  ``Q.integer(x)`` is true iff ``x`` belongs to the set of integer
 |  numbers.
 |  
 |  Examples
 |  ========
 |  
 |  >>> from sympy import Q, ask, S
 |  >>> ask(Q.integer(5))
 |  True
 |  >>> ask(Q.integer(S(1)/2))
 |  False
 |  
 |  References
 |  ==========
 |  
 |  .. [1] https://en.wikipedia.org/wiki/Integer
 |  
 |  Handler
 |  =======
 |  
 |  Multiply dispatched method: IntegerHandler
 |  
 |  Handler for Q.integer.
 |  
 |  Test that an expression belongs to the field of integer numbers.
 |  
 |  Inputs: <Add>    
 |      * Integer + Integer       -> Integer
 |      * Integer + !Integer      -> !Integer
 |      * !Integer + !Integer -> ?
 |      
 |  
 |  Inputs: <Pow>    
 |      * Integer + Integer       -> Integer
 |      * Integer + !Integer      -> !Integer
 |      * !Integer + !Integer -> ?
 |      
 |  
 |  Inputs: <Mul>    
 |      * Integer*Integer      -> Integer
 |      * Integer*Irrational   -> !Integer
 |      * Odd/Even             -> !Integer
 |      * Integer*Rational     -> ?
 |      
 |  
 |  Other signatures:
 |      * Expr
 |      * Rational
 |      * Infinity
 |      * NegativeInfinity
 |      * GoldenRatio
 |      * Integer
 |      * Pi
 |      * TribonacciConstant
 ...
```

And here is how it's displayed in sphinx docs.

![image](https://user-images.githubusercontent.com/52185322/104160256-2cb4a180-5434-11eb-8869-1d95dbcb5645.png)

JSS95 committed:
Modify __doc__ and enhance docs

oscarbenjamin commented:
Looks good to me. I'll merge this now
